docs: document gcp backup (#127)
rust-lang/infra-team: service-catalog/gcp-backup/README.md

> # GCP backup

Reviewer: I think GCP is an implementation detail and would therefore rename this piece of our infrastructure (i.e. the document). Internally, we've talked about this in terms of "backup of critical (Rust) assets". How about we use a similar name for this "service"?

## Summary

In GCP (Google Cloud Platform) we keep offsite backups for both Rust releases and crates
to protect us against security threats that could involve losing crates or releases.

> These threats were identified in the [threat model].

Reviewer: ```suggestion
These threats were identified in a [threat model] for the project's infrastructure, created by the Rust Foundation's security engineer Walter.
```

There are multiple threat models, and I think it's be nice to call out @walterhpearce who created them. 🙂

## Motivation

While we have multiple measures in place to prevent accidental deletion of Rust releases or crates in AWS,
e.g. bucket replication to a different region and restricted access, our current setup does not sufficiently protect us against a few threats:

> 1. _AWS Account compromise_. The [threat model] for Rust's infrastructure, created by the Rust Foundation's security engineer, highlights the risk of an AWS account compromise.

Reviewer: ```suggestion
1. _AWS Account compromise_. The [threat model] highlights the risk of an AWS account compromise.
```

I suggest that we add the attribution above and then just keep referencing the threat model here.

   If a malicious actor was able to gain administrator access to the AWS account of one of the [infra-admins],
   they could bypass a lot of safe guards and delete data.
2. _AWS Account deletion_. AWS could accidentally delete our account, resulting in the possible deletion of data and backups.

>    For example, something similar happened at [Google](https://arstechnica.com/gadgets/2024/05/google-cloud-accidentally-nukes-customer-account-causes-two-weeks-of-downtime/) recently.

Reviewer: ```suggestion
   Something similar happened to a customer on [GCP](https://arstechnica.com/gadgets/2024/05/google-cloud-accidentally-nukes-customer-account-causes-two-weeks-of-downtime/) in 2024.
```

In a few months, we won't know what "recently" means. 😅

- To mitigate threat 1, the new backup needs to have separate admin access.
- To mitigate threat 2, the new backup needs to be in a separate cloud environment.

## Implementation overview

These new backups are hosted in a dedicated GCP account and have totally separate access controls compared to AWS.
Specifically, none of the current `infra-admins` have admin access to this separate environment to protect against an account compromise.
This GCP account is not used for anything else (just for backups).

The backups are automatically copied daily by GCP.

### Access 👤

We limit admin access to the GCP backups to two members of the Rust Foundation for the following reasons:

- _ensure a strong separation of access_: as explained in the first [motivation](#Motivation), the GCP admins should be different from the AWS admins.
  This means we can't give admin access to any of the `infra-admins`.

> - _accountability_: the Rust Foundation employees are accountable for their actions, which means they can't delete the backups intentionally without breaking the law.

Reviewer: ```suggestion
- _accountability_: The Rust Foundation employees have signed an employment contract and can be legally liable for malicious actions.
```

People with admin access to the GCP account:

- Joel Marcey (Director of Technology @ Rust Foundation)
- Walter Pearce (Security Engineer @ Rust Foundation)

People with read-only access to the GCP project:

- `infra-admins`

The admin access of the GCP account is bound to the `@rustfoundation.org` Google Workspace account.
This means that if an employee leaves the Rust Foundation, they lose access to the GCP account.
In this case we need to add a new admin to the GCP account.

> [!NOTE]
> The `infra-admins` team can have admin access to the GCP staging project if needed.

### In case of emergency 🧯

- In case our data in AWS is deleted, the `infra-admin` team can restore it by:
  - copying the data from GCP to AWS using the GCP read-only access.
  - restoring the `crates-io-index` bucket from the `db-dump` stored in the `crates-io` bucket. Use [this](https://github.com/rust-lang/crates.io/blob/e0bb0049daa12f5362def463b04febd6c036d315/src/worker/jobs/git.rs#L19-L129) code.

> - If the GCP synchronization mechanism breaks, the Infra team can raise a PR to fix the terraform configuration and a GCP admin can apply it.

Reviewer: Nit: I'd either call the the "(I|i)nfrastructure team" (without a strong preference for capitalization) or the "infra-team", with a preference for the first. We're optimizing for readability here, and imo abbreviations make it more difficult to read long documents. But that's totally a personal opinion of mine...

Author: [82e3135](https://github.com/rust-lang/infra-team/pull/127/commits/82e31351f76a41b7ae3b5646a36ca26f3105ee53) 👍

> - If the GCP synchronization mechanism breaks, the Infra team can raise a PR to fix the terraform configuration and a GCP admin can apply it.

Reviewer: Nit (another personal opinion): When talking about Terraform, the product, we should capitalize its name. When talking about `terraform`, the CLI, we should wrap it in a code "block". (This goes for all product names, e.g. CloudFront further down.)

Author: [1000b25](https://github.com/rust-lang/infra-team/pull/127/commits/1000b25fe80c7d3884704d9a339ad80538b3968f) 👍

- one GCP admin account (Joel or Walter)

This improves our security posture because compromising two accounts is harder than compromising one.

The accidental account deletion is not a threat anymore because if either AWS or GCP delete our account, we can restore the data from the other provider.

## Implementation details

The account where we store the backup is called `rust-backup`. It contains two GCP projects: `backup-prod` and `backup-staging`.
Here we have one Google [Object Storage](https://cloud.google.com/storage?hl=en) in the `europe-west1` (Belgium) region for the following AWS S3 buckets:

- `crates-io`. Cloudfront URL: `cloudfront-static.crates.io`. It contains the crates published by the Rust community.
- `static-rust-lang-org`. Cloudfront Url: `cloudfront-static.rust-lang.org`. Among other things, it contains the Rust releases.

For the objects:

- Set the [storage class](https://cloud.google.com/storage/docs/storage-classes) to "archive" for all buckets.
  This is the cheapest class for infrequent access.
- Enable [object-versioning](https://cloud.google.com/storage/docs/object-versioning) and [soft-delete](https://cloud.google.com/storage/docs/soft-delete),

>   so that we can recover updates and deletes so that we can recover updates and deletes.

Reviewer: ```suggestion
  so that we can recover updates and deletes.
```
